src/components/tl/ucp/tl_ucp_coll.c

```diff
@@ -24,14 +24,58 @@
 #include "fanout/fanout.h"
 #include "scatterv/scatterv.h"
 
-const char
-    *ucc_tl_ucp_default_alg_select_str[UCC_TL_UCP_N_DEFAULT_ALG_SELECT_STR] = {
-        UCC_TL_UCP_ALLGATHER_DEFAULT_ALG_SELECT_STR,
-        UCC_TL_UCP_ALLREDUCE_DEFAULT_ALG_SELECT_STR,
-        UCC_TL_UCP_BCAST_DEFAULT_ALG_SELECT_STR,
-        UCC_TL_UCP_ALLTOALL_DEFAULT_ALG_SELECT_STR,
-        UCC_TL_UCP_REDUCE_SCATTER_DEFAULT_ALG_SELECT_STR,
-        UCC_TL_UCP_REDUCE_SCATTERV_DEFAULT_ALG_SELECT_STR};
+ucc_status_t ucc_tl_ucp_team_default_score_str_alloc(ucc_tl_ucp_team_t *team,
+    char *default_select_str[UCC_TL_UCP_N_DEFAULT_ALG_SELECT_STR])
```

Comment:
I would just maybe tried to compact this fn a bit to avoid code dup.
Maybe, we can define a struct :
```
typedef char* (*score_str_get_fn_t)(ucc_tl_ucp_team_t *team);
typedef struct ucc_tl_ucp_default_alg_desc {
    char *select_str;
   score_str_get_fn_t str_get_fn;
} ucc_tl_ucp_default_alg_desc_t;
```
Then each component that currently defiens either STR or FN will actually define a struct where only 1 entry can be non-null. And we will have the same array:
```
const ucc_tl_ucp_default_alg_desc_t
    ucc_tl_ucp_default_alg_descs[UCC_TL_UCP_N_DEFAULT_ALG_SELECT_STR] = {
        UCC_TL_UCP_ALLGATHER_DEFAULT_ALG_SELECT_STR,
        UCC_TL_UCP_ALLREDUCE_DEFAULT_ALG_SELECT_STR,
        UCC_TL_UCP_BCAST_DEFAULT_ALG_SELECT_STR,
        UCC_TL_UCP_ALLTOALL_DEFAULT_ALG_SELECT_STR,
        UCC_TL_UCP_REDUCE_SCATTER_DEFAULT_ALG_SELECT_STR,
        UCC_TL_UCP_REDUCE_SCATTERV_DEFAULT_ALG_SELECT_STR};
```
but now with descs instead of strings.
Then in the fn below you will have just a single loop over all descs, and you will select either strdup if select_str != nUll or call fn otherwise.
wdyt?

Reply:
looks good to me, added